### DISCUSSION TITLE
Recalculating clinical treatment plans using DOSXYZnrc

### DISCUSSION THREAD

**Discussion by marenaud** [nrc-cnrc/EGSnrc#849] [General]
When using particle sources that interpolate beam parameters based on a sampled `muIndex` (e.g sources 20, 21),
there is a "gotcha" that you should be aware of.

An interpolation-related issue can occur when any of the angular values crosses 360°. For example, suppose we have the following control points (with only the relevant details provided):

    ...
    cpt N-1: phi = 348°, weight = 0.4
    cpt N:   phi = 355°, weight = 0.5
    cpt N+1: phi = 2°, weight = 0.6
    cpt N+2: phi = 9°, weight = 0.7
    ...

If, for a given particle history, we get a randomly sampled `muIndex = 0.55`, then we may naively expect to be halfway between control points N and N+1, at `phi = 358.5°`.

However, from equation 1 on page 64 of the [DOSXYZnrc manual](https://nrc-cnrc.github.io/EGSnrc/doc/pirs794-dosxyznrc.pdf) or [this place in the code](https://github.com/nrc-cnrc/EGSnrc/blob/master/HEN_HOUSE/user_codes/dosxyznrc/srcxyznrc.mortran#L4195), we can calculate that the interpolated phi value will be `355 + (2 - 355) / 2 = 178.5°` which is quite different from what we expected! This situation can easily happen when programmatically converting control points from a DICOM RTPlan into DOSXYZnrc control points, and is especially common in Tomotherapy treatment plans as the gantry does multiple rotations around the patient.

One of the simple workarounds consists of extending the domain to infinity when parsing the treatment plan. Everytime we cross the 360° boundary, we increase a `num_revolutions` counter and add `num_revolutions * 360°`
to the angular values of each control point, so that the angular values go from `0..360..720..1080..` and so on. Another workaround would be to insert duplicate control points at 360 and 0 with the same weight every time an angular component loops around.

If others have dealt with this "issue", feel free to share your thoughts on the workarounds!

**Comment by ojalaj** [comment]
@marenaud  Wow! Did not see this happening. How did you come up with this? Have you seen any visual and/or dosimetric effects of this issue? With conventional C-arm linac treatment plans with multiple arcs? Or only with Tomotherapy plans?

**Reply by marenaud** [reply]
I had originally thought about this issue back in 2011 when making a tomotherapy model which required coding new synchronised sources between BEAM and DOSXYZ because sources 20 and 21 and the SYNC CMs didn't exist. Then I completely forgot about it when reimplementing everything using the new sources.

The effect is more obvious for tomotherapy plans because control points span 7 degrees, so you get a non negligible chunk of fluence that gets BEAMed from the wrong angle, but it's typically a minor blip attributed to MC uncertainty.

Since VMAT plans tend to have much tighter control points (on the order of 1deg) the affected regions will only become a problem if a ton of fluence is being dumped through those specific control points. It can actually happen without doing a full arc around the patient, since the crossing between `phi = 360` and `phi = 0` happens at gantry angle of 90 for a head-first supine coplanar delivery when using the transformations from Zhan et al, 2012 "Beam coordinate transformations from DICOM to DOSXYZnrc".

I can imagine it also being potentially problematic for non-coplanar deliveries since both `theta` and `phicol` will be changing a lot as well, as opposed to coplanar deliveries where `theta = 90` usually and `phi` is the only degree of freedom that changes. If I were recalculating those, I'd probably go through the trouble of duplicating control points and keeping the domain to [0, 360].

/u/Paintasaurus reports also seeing some effects on reddit (https://www.reddit.com/r/EGSnrc/comments/stf2jx/public_service_announcement_for_people/hx9xpz9/)

**Comment by ftessier** [comment]
Good find, and thank you @marenaud for taking the time to point this out. If I understand correctly, this is not something we ought to fix in DOSXYZnrc, in the sense that one _could_ in principle want a control point at 355°, and then interpolating back to 2°, around the nearly full circle (as opposed to 362°). Perhaps the code could issue a warning when the control points are not increasing or decreasing monotonically?

**Reply by marenaud** [reply]
I agree, it's not actually a bug and there's a legitimate reason for someone to go from 355 to 2 counter-clockwise so they'd be pretty confused if there were errors!

It's just an easy-to-overlook issue since often these control points are one among hundreds / thousands being created algorithmically. We could also think of compiling a community checklist of common issues that people should keep in mind when recalculating clinical plans or generating control points programmatically.

**Reply by ftessier** [reply]
Good idea! Go ahead and start a new Discussion thread! 👍🏻

**Reply by ftessier** [reply]
Done, see #850, also pinned at top of Discussions.

**Reply by marenaud** [reply]
Merci @ftessier, I'll try to contribute when I can find some free time :) There's a few topics that would probably benefit from community discussion and agreeing on best practices, like highlighting a few papers published on converting DICOM angles to DOSXYZ angles and taking into account different patient orientations

**Reply by ftessier** [reply]
Indeed, this could become a very useful resource (and eventually lead to a review paper, why not?). I suggest starting topic threads, as was done here, and link to the individual discussions from the checklist, to keep the it uncluttered.

**Comment by crcrewso** [comment]
A thought, In the RTPlan there is `Gantry Rotation Direction (300A,011F)` could we use this in the conversion to automatically insert  360 and 0 control points.

**Reply by marenaud** [reply]
We'd definitely have to keep the rotation direction in mind, though now I wonder about plans where the gantry would change direction. I guess each change in direction would add a new Beam in the DICOM's BeamSequence so it would be okay. One tricky bit is to keep in mind that the interpolation "issue" happens when the DOSXYZ angular degrees of freedom loop over, not the DICOM gantry / couch / col angles, so the crossover points will be dependant on coordinate transforms and patient position

**Comment by ojalaj** [comment]
Dear EGSperts (especially the ones contributing to the conversatino above - @marenaud, @ftessier, @crcrewso, but also @rtownson @blakewalters).

With source=20, simulating "normal" C-arm linac, phantom HFS, I have plan with five fields - four static fields with zero gantry and collimator angle etc. and then I have one half arc (from 0 to 180, clockwise), which is the second field, I got some weird results.

To me it seems that the half arc has been simulated from 180 --> 360 CW or 0-->180 CCW (not sure which way since the MLCs are static all the time, and thus it doesn't matter), i.e., wrong side of the phantom. I see the dose distribution mirrored.

Did I bump into this issue reported here?:
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0000
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0003
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0003 -------------------this is the half arc
0.000, 0.000, 0.000, 90, 90.00, 270, 100.0, 0.5000---------------------this is the half arc
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.5000
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.5003
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.5003
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7502
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7502
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 1.0000

If yes, how to deal with this? Manually changing the field (and jaw/MLC sequence) order so that the phi would be increasing, i.e. inverting the half arc sequence and moving it to first?

If not, what is going on here :) ?

**Reply by marenaud** [reply]
Hi Jarrko!

That definitely looks like a good example for an unexpected result from this issue :)

Since these are static fields, you could add 2 control points to the dosxyz file

```
0.000, 0.000, 0.000, 90, 360.00, 270, 100.0, 0.25
0.000, 0.000, 0.000, 90, 0, 270, 100.0, 0.25
````

That'll fix the interpolation to always have the right reference. Alternatively make phi go from 270 to 450 instead of 90, ie go outside the usual 2pi domain.

**Reply by ojalaj** [reply]
Hello Marc-André!

Thanks for prompt reply.

I apologize for dumb question, but if I add those control points to my .egsinp file, do I also need to add additional, corresponding (sub-)sequences (duplicates) to my jaw and MLC sequence files?  So instead of 10 control points I would have 12 control points, but still 10 (sub-)sequences or 12 (sub-)sequences?

Or could I alternatively just replace this one
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0003 -------------------this is the half arc
0.000, 0.000, 0.000, 90, **_90.00_**, 270, 100.0, 0.5000---------------------this is the half arc
with
0.000, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0003 -------------------this is the half arc
0.000, 0.000, 0.000, 90, **_450.00_**, 270, 100.0, 0.5000---------------------this is the half arc
?

I would be very very happy if this could be handled somehow at code level or if there at least would be some kind of warning that the user should pay attention to some specific lines.

**Reply by marenaud** [reply]
I think here you could get away with decoupling the beamnrc and dosxyz control points so you'd only need to add them to the dosxyz file.

But the alternative of replacing 90 with 450 is the probably simplest option since it doesn't involve adding any control points. That's what I do as well.

**Reply by ojalaj** [reply]
And can the rest of the control points/angles remain as they are, i.e., after changing 90 --> 450, the rest of the phis don't need to be increasing/larger?

**Reply by marenaud** [reply]
I think with your current arrangement of control points, you won't need to change the others. The interpolation function will grab the first control point with an index higher than the randomly generated number, and interpolate between that one and the previous one. By having another control point at 270 with the same index as the one at 450, you'll ensure that any muIndex generated above 0.5 will be interpolated from 270 instead of 450. And of course the rest of your control points are all at 270 so basically nothing weird should be happening for those.

**Reply by TonyPopescu** [reply]
Hi everyone,

Thanks for this useful discussion and for bringing up this important topic. I agree with Frédéric that this is not a "bug", so we should not address this at the level of the DOSXYZnrc code, but at the level of our respective local codes that we use to generate the egsinp files, based on RTP parameters. (Locally, we call these 'DICOM_2_egsinp' codes.) 

I clearly remember that back in the days of my work with Julio Lobo, we identified this potential interpolation error and made our DICOM_2_egsinp automatically insert the phi = 360 & 0 control points at the appropriate (same) MU index. However, prompted by the present discussion, I revisited the problem and discovered that we are also vulnerable to this error. We did not notice it, because virtually all of our simulations are for VMAT plans, for which the error is small. Over many years, our set of codes evolved several times, were taken apart and reassembled, and somehow the old 'catch' disappeared from the current version. Sadly, in the meantime I have shared my codes with many people, so I will have to do a 'worldwide recall' after I fix this. Also, as was mentioned above, it is important to rely on the CW/CCW information from the RTP file, in order to not second guess if the insertion is needed.

Quick reply to one of Jarkko's questions: the number of control points can be as different as needed between the DOSXYZnrc egsinp, the jaws sequence file, and the MLC sequence file. That will not impact the synchronization.

I really like the idea of a review paper, with contributions from all present and others, on the current (best) practices to generate egsinp files, deal with coordinate transformations, etc., especially in challenging situations, overview of the various ways people have used sources 20 and 21 and, especially, have all of us contribute to a repository of such 'DICOM_2_egsinp' codes, ideally as a branch of the EGSnrc github. I think the use of these sources (and, consequently, the MC QA of clinical plans) would be more widespread if the community had this kind of support.

**Reply by rtownson** [reply]
Hi Tony,

I think the review paper or at least contributions of relevant utility codes would be very helpful to the community. We have started looking into including third-party contributions as git submodules. I think this lets us keep those contributions a bit separate and optional to users, while also encouraging upkeep by the original authors. If you're interested in including DICOM_2_egsinp, that would be great.

**Comment by ojalaj** [comment]
EDIT: Since I don't know do we get notified of the Discussions similarly than of Issues and PRs, I'll add @marenaud @TonyPopescu @ftessier @crcrewso @rtownson 

Did I understand correctly from above that this issue can be generalized to all variables theta, phi and phicol, i.e. if I have four static MLC fields, but the second field has a different collimator and couch angle, there is a potential problem in below, when moving from second field to third one?:

0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0000
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, **325**, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, **270**, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7981
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7981
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 1.0000

If yes, can this be handled by adding?:
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0000
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.5962
**0.040, 0.000, 0.000, 90, 270.00, 360, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, 0, 100.0, 0.5962**
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7981
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7981
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 1.0000

**Reply by rtownson** [reply]
I haven't looked in detail, but my understanding from the discussion above was that you only need to worry when crossing over the 0/360 boundary *during a change in MU*. In that case, you need to add two new control points with equal MU at 360 and 0, or the reverse for CCW. 

Or, to avoid interpolating the MU to 360, you could use 360+the angle of the next control point (e.g. 450 for one of the examples above), and then you would also want to add 360 to every following control point until a static change. 

Or, instead of adding 360 to every following control point, you could add one new control point after 450, at 90 with the same MU. This ensures that the *next* control point is interpolated between its angle and 90, not 450. 

For the above, since these are static fields and you are passing 360 *without a change in MU* (e.g. 325 at 0.5962 and 270 at 0.5962), then you have nothing to worry about.

@marenaud is my understanding correct?

**Reply by marenaud** [reply]
That's right! It's only an issue for the dynamic components of any input file, and it does indeed affect any of the angular components.

@ojalaj Those control points are purely static so there's no issue there. In your previous example the only issue was with your half-arc. If you refer back to the formulas my first post, you'll see the reason why: it all has to do with how angles are interpolated between successive control points. Static deliveries have their weights set up such that no interpolation ever happens.

If the sampling routine generates a muIndex between 0.5962 and 0.7981, the control points in bold below will be the ones used to calculate beam characteristics
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0000
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.5962
**0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7981**
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.7981
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 1.0000

So you can see that there's no interpolation issues and no need to add extra control points.

In fact, to reinforce the point that Tony made in his post above about how dosxyznrc control points are decoupled from BEAMnrc control points, your control points above could actually be simplified down to
```
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0000
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.0050
0.040, 0.000, 0.000, 90, 270.00, 325, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 0.5962
0.040, 0.000, 0.000, 90, 270.00, 270, 100.0, 1.0000
```
with the same results.
